### ISSUE TITLE
The release task needs to carefully choose packages to release

### ISSUE THREAD

**Issue by Reinmar** [ckeditor/ckeditor5-dev#64]
Right now it's gonna release every single package (dependency of the main repo). It needs to be way smarter and release packages which:

1. Had any changes (even internal ones – after all, it will be just a patch release). The idea is that we'd generally want to have  latest releases on npm, even if only things like docs have changed. Otherwise, we'd risk having e.g. an outdated README on npm for months (if not years – see e.g. how few commits to the code were made in https://github.com/ckeditor/ckeditor-dev/commits/master/plugins/basicstyles/plugin.js). Also, we may mistakenly miss documenting some change, so really – it's just safer to release it any commit was made.
2. If no commit was made, release a package if any of its dependencies (excluding devDependencies!) will be released (and needs to be updated in `package.json`). The idea, in this case, is that if something got fixed in a dependency, it also might affect the dependent package. It's also quite clear I think that it always needs to be a patch release because new features added to dependency do not affect the dependent package. However, a minor or major release of a dependency might bring patches so a depedent package needs to be bumped up too.

The last point doesn't seem very hard to implement, but it is... because currently the changelog task creates version entries in packages which haven't changed. So it must work a bit differently. It must not create new version entry for a package with no commits and this entry must be created by the release tool if it release the package anyway.

**Comment by pomek** [comment]
TL;DR: I implemented the idea above. It even works but incorrectly and I'm considering to rewrite the whole mechanism.

But I will start at the beginning.

I created a function - [`hasCommitsUntilLastRelease()`](https://github.com/ckeditor/ckeditor5-dev/blob/t/64/packages/ckeditor5-dev-env/lib/release-tools/utils/hascommitsuntillastrelease.js) which tells me whether the package has commits from the last tag. If the package does not have any tag, the function will check whether any commit was made. It is very useful for both tasks.

### Changelog

The function is used by a task for generating the changelog. The task checks [whether the changelog can be generated](https://github.com/ckeditor/ckeditor5-dev/blob/t/64/packages/ckeditor5-dev-env/lib/release-tools/tasks/generatechangelog.js#L28-L32). If no, a process is aborted.

It works properly. Changelog won't be generated when any commit wasn't made.

### Release

Unfortunately, a task for creating new release is more complicated.

We can release a package in two cases:

1. The package has commits. We don't make empty commits so every commit means some change.
2. The package does not have new commits but… dependencies of the package may change (dependencies can have committed changes). In this case, we should release the package with a new `patch` version (following the [SemVer](http://semver.org/)).

* In all other cases, the package must not be released.

The scenario above is easy to implement:

1. [The package has new changes. Let release it.](https://github.com/ckeditor/ckeditor5-dev/blob/t/64/packages/ckeditor5-dev-env/lib/release-tools/tasks/createrelease.js#L39)
2. [The package's dependencies can change themselves.](https://github.com/ckeditor/ckeditor5-dev/blob/t/64/packages/ckeditor5-dev-env/lib/release-tools/tasks/createrelease.js#L46-L63) This part of the code tries to update dependencies of the package. Then the changes are committed. After that, changelog can be created (the package has new change) and the whole release process can be repeated. We will go the first point.

* [There are no changes, so the whole process is aborted.](https://github.com/ckeditor/ckeditor5-dev/blob/t/64/packages/ckeditor5-dev-env/lib/release-tools/tasks/createrelease.js#L41-L46)

So, what is the problem?

Let's assume we have repositories (and their dependencies) like a tree below:

```
- ckeditor5 # does not have changes
  |- ckeditor5-engine # does not have changes

- ckeditor5-engine
  |- ckeditor5-core # has changes
```

The release process described above should:

1. Release a package `ckeditor5-core` because it contains new changes.
2. Release a package `ckeditor5-engine` because dependencies of the package have changed. `ckeditor5-core` will release a new version.
3. Release a package `ckeditor5` because dependencies of the package have changed. `ckeditor5-engine` will release a new version.

Are the assumptions correct?

At this moment, the process is impossible to do. Why? 

We need to know how the release process works now.

1. For each a package, the mechanism tries to predict the new version for the package. If the package does not have changes - the process is aborted.
2. The mechanism knows versions of the dependencies. But the dependencies can be incorrect. The mechanism based on the commits and does not know relations between each other. For the tree above, `ckeditor5-engine` and `ckeditor5` won't have a new version.
3. Mechanism will release `ckeditor5-core`. The rest packages will be aborted.

The process "works" but not properly. The release task released a package with changes (`ckeditor5-core`) but the related packages were skipped.


## Solution

So what is the solution?

I'm afraid we need to rewrite the mechanism.

#### But before we start

A function [`executeOnDependencies()`](https://github.com/ckeditor/ckeditor5-dev/blob/t/65/packages/ckeditor5-dev-env/lib/release-tools/utils/executeondependencies.js) is great. It allows executing a custom function for each dependency. But, what if we want to skip some package? The `handler` needs to know which package should be aborted. It makes the function more unreadable.

Let introduce a new function (or modify the existing one) - `executeOnDirectories( /* Array.<String> */ directories, /* Function */ handler )`. This function will execute the `handler` only for given `directories`. It allows controlling where we want to execute the action.

### Changelog

1. Let's assume that for all dependencies will be generating the changelog. 
2. For each package, we need to check whether the changelog should be generated. `hasCommitsUntilLastRelease()` will help.
3. After **2.** we will have dependencies with changes. For these packages, we can generate the changelog.
4. For each package, a user has to confirm the version. The user can also change the version manually if he wants.
5. We have dependencies with confirmed versions - let's generate the changelog.

Changes in the actual version:

- Filters out dependencies before generating the changelog.
- The user has to confirm proposed version.

### Release

1. Let's assume that for all dependencies will be creating the release. The same as in the changelog task.
2. For each package, we need to check whether the release should be created. Rules are the same like above.
3. Somehow we need to filter all packages to release. A process is described below.
4. Packages are ready to release. Next version must be taken from the changelog file.
5. Release packages.

Changes in the actual version:

- Rewrite all the process.

#### How to filter the packages?

As the example I will extend the tree above:

```
- ckeditor5 # does not have changes
  |- ckeditor5-engine # does not have changes

- ckeditor5-engine
  |- ckeditor5-core # has changes
  
- ckeditor5-basic-styles # does not have changes and dependencies
```

The process:

```
# First iteration. 
- check: `ckeditor5` - result: no; no changes in changelog
- check: `ckeditor5-core` - result: yes; changes in the package
- check: `ckeditor5-engine` - result: no; no changes in changelog
- check: `ckeditor5-basic-styles` - result: no; no changes in changelog

// All "no" from the first iteration are added to a queue.

# Second iteration (elements from the queue):
- check `ckeditor5` - result: no; no changes and dependencies won't changes; increase the counter (1)
- check `ckeditor5-engine` - result: yes; dependencies have been updated (`ckeditor5-core`)
- check `ckeditor5-basic-styles` - result: no; no changes and dependencies won't changes; increase the counter (1);

// All "no" fom the second iteration are parsed one more time. Because at least one package will be released, we need to reset the counter for all packages.

# Third iteration:
- check `ckeditor5` - result: yes; dependencies have been updated (`ckeditor5-engine`).
- check `ckeditor5-basic-styles` - result: no; still no changes; increase the counter (1)

# Fourth iteration:
- check `ckeditor5-basic-styles` - result: no; still no changes; increase the counter (1)

# Fifth iteration:
- check `ckeditor5-basic-styles` - result: no; still no changes; increase the counter (2) and remove from the queue.
```

We don't know how many iterations we need to filter out invalid packages. 

Packages are removed from the queue if the counter is equal to 2. It means, the mechanism checked two times the package and results of the checking are the same. Dependencies won't change.

For `ckeditor5-engine` and `ckeditor5` the mechanism needs to update version of dependencies in `package.json` and commit the changes. After that, we are able to generate changelog. The user must not confirm the version. **If we update dependencies only, we always bump the `patch` version.**


## Validation

Before the generating changelog or creating the release, we need to validate repository state.

#### Changelog

1. Is current branch a master?
2. Is the master up to date?
3. Is the master ahead the repository?

#### Release

1. The same as for `changelog`.
2. Does user pass the `token`?

**Comment by Reinmar** [comment]
I'll first comment on particular aspects and then I'll propose necessary changes.

> Filters out dependencies before generating the changelog.

I'm not sure I understand this. You mean that a changelog won't be generated if only dev deps have changed? How are you going to check that?

BTW, we need to update dev deps from time to time too. In fact, the release task must do it either  **for all packages** or before releasing a specific package if it has to be released (due to changes or normal deps updates). This complicates the picture a bit more.

> The user has to confirm proposed version.

And what if the versions are wrong? We need an option to proceed somehow. Otherwise, it will be a stalemate. So the releaser must be able to accept the proposed version or choose a version by himself/herself. The releaser must be asked about **every package**, not just the changed ones. This is required so we're able to release `1.0.0-alpha.1`, then `1.0.0-alpha.2`, etc. In this case, the task is not able to make a decision itself and in some of these cases the releaser must be able to release every package, even unchanged.

We also need to make sure that the release task won't break on alpha/beta/rc versions.

> Packages are removed from the queue if the counter is equal to 2. It means, the mechanism checked two times the package and results of the checking are the same. Dependencies won't change.

This isn't true, I'm afraid. Waiting 2 iterations may not be enough if a chain of deps is longer.

---

I'm seeing that this task becomes more and more complex with every new requirement and the number of requirements makes it really tricky, so we must be very careful.

**Comment by Reinmar** [comment]
I'm posting here a draft of the process. I'll work on the next steps on Monday.

## Changelog task

1. For every packages:
   1. Check if any code changes were done.
      1. If so, propose a version bump (note: make sure to handle alpha/beta/rc releases properly).
      2. If no, propose skipping.
   2. Let the user decide whether the tool is right and correct it if it's mistaken. In other words, the user should be able to propose a new version (by typing it). If the user chose to release unchanged package, generate an empty changelog entry for it. If the user chose not to release a package, do nothing.
   3. Add changelog entries for a picked version and commit the changelog. Do nothing if the package is meant to be skipped.

This should be it.

## Release task

A lot more complicated job, but let's make sure to split the process into clear steps.

### Building a package version database

This step must be done before anything else happens. It will make it testable for us now and for the releaser later on.

TODO??? Let's assume here that changes to dev dependencies will be made in separate commits marked with a special type "Internal (devdeps)". This will allow easilly validating whether (TODO: I must think on this)

1. Get proposed version for every package from its changelog. Yes, from its changelog, so we use exactly the version which the releaser had chosen when running the changelog task.
2. Match those versions with the latest tag to filter out packages which are not supposed to be released.
3. At this point, we have a list of packages which must be released.
4. Now, find dependent packages:
   1. Create a map of all packages. Mark those to be released (from point 2.).
   2. Set `cleanRun` to `false`.
   3. While `cleanRun === false`:
      1. Set `cleanRun` to `true`.
      2. Fetch packages to be released from the map.
      3. Fetch packages to not be released from the map.
      4. Iterate over packages to not be released and update their deps based on the map of packages to be released. If any change has to be done set `cleanRun` to `false.

At this point we should have a map of packages to release and those to skip. Print both to the user and ask to confirm. Ask only to accept or reject the whole list. If any package is missing, the releaser can add a changelog entry in that package (because it means that it wasn't generated by the changelog task or that there's some problem with tags) and repeat the proces.

### Next

TODO...

**Comment by pomek** [comment]
Could you provide an example which will show how your algorithm works?

**Comment by Reinmar** [comment]
> Could you provide an example which will show how your algorithm works?

It's ~20-40LOC to implement a test version of this algorithm. I think you'll get more from that. It's a really simple algorithm.

**Comment by pomek** [comment]
> If no commit was made, release a package if any of its dependencies (excluding devDependencies!) will be released

What if any of dependency has bumped the `major` version?

Example:

```
ckeditor5-core # no changes, deps: engine, v1.0.0
ckeditor5-engine # bump from 1.0.0 to 2.0.0 (BC)
```

Based on our assumptions - version of `ckeditor5-core` should be bumped because dependencies have changed.

```
ckeditor5-core #bump deps new version: v1.0.1
```

`ckeditor5-core` won't work properly because `ckeditor5-engine` has changed its API.

**Comment by Reinmar** [comment]
> What if any of dependency has bumped the major version?

I'm not sure I understand. The same rule applies – the dependent package must be released. I even explained that in the original post:

> However, a minor or major release of a dependency might bring patches so a depedent package needs to be bumped up too.

So it doesn't matter if a dependency was bumped up by a patch, minor or major. Only the fact that a bug might've been fixed in it requires releasing the dependent package.

You're asking what if a dependency changed its API. There are two options then:

* those changes required aligning the dependent package, so it contains commits (otherwise, tests wouldn't pass),
* those changes didn't require any change in the dependent package and then we need to release it because the bug fixed in the dependency might also affect the dependent package.

However, I'm thinking now that perhaps this is a too idealised picture. It doesn't consider the fact that API changes in a dependency might change the API of the dependent package too. Theoretically, dependent package's API should not expose its dependencies, but this is the perfect design which is never true.

So, perhaps the rule should be that:

* patch or minor bump of a dependency should result in at least a patch bump of a dependent package,
* major bump of a dependency should result in a major bump of a dependent package.

But that'd mean that a major release of the engine will mean a major release of any other package. That doesn't make sense, does it?

**Comment by Comandeer** [comment]
Probably it requires one more step between releasing new package version and releasing new dependent packages versions: testing and checking if there are some breaking changes. If there are, then the tool should skip that package and mark that its next release should be major. If not, just publish the patch version. Something similar to what [cracks](https://github.com/semantic-release/cracks) does.

**Comment by Reinmar** [comment]
Hm... Cracks doesn't sound like a viable solution to me. Maybe I'm wrong, but relying on your test suite in order to tell you whether there've been any breaking changes assumes that the test suite is exceptionally well designed and complete. I'd rather rely on the humans here ;>.

Besides, we're talking here about the situation when a package X hasn't changed at all and package Y, on which X relies, has changed. X's test suite hasn't changed either, so I'd only like to know how the new version of Y affects it and X's tests suite should not check e.g. how the image's balloon panel (which comes from ckeditor5-ui, but is available by reference in ckeditor5-image) works. There are ckeditor5-ui's tests to verify that.

Theoretically, this should be true:

> Theoretically, dependent package's API should not expose its dependencies, but this is the perfect design which is never true.

Hence, we've got the changelog entries in packages which changed and those entries gives us a hint how they changed. Based on that knowledge we need to decide how to release the dependent packages.

> * major bump of a dependency should result in a major bump of a dependent package.
>
> But that'd mean that a major release of the engine will mean a major release of any other package. That doesn't make sense, does it?

Looking at this,
